PyDynamic/uncertainty/propagate_filter.py

```diff
@@ -167,99 +168,274 @@ def FIRuncFilter(y, sigma_noise, theta, Utheta=None, shift=0, blow=None, kind="c
+            P = A @ P @ A.T + np.square(Ux[n]) * np.outer(bs, bs)   # state uncertainty, formula (18)
+        else:  # "corr"
+            P = A @ P @ A.T + Ux[0] * np.outer(bs, bs)              # state uncertainty, adopted from formula (18)
+        
+        #dA_z = np.hstack(dA @ z)                                   # not efficient (dA is sparse), dA = _get_derivative_A(p)
+        dA_z = np.vstack((np.zeros((p-1, p)), -z[::-1].T))          # efficient, no tensor-multiplication involved
+        dz = A @ dz + dA_z                                          # state derivative, formula (17)
+        z = A @ z + bs * x[n]                                       # state, formula (6)
+        
+    Uy = np.sqrt(np.abs(Uy))  # calculate point-wise standard uncertainties
 
-    Aabl = np.zeros((p, p, p))
-    for k in range(p):
-        Aabl[0, k, k] = -1
+    # return result and internal state
+    state.update({"z": z, "dz": dz, "P": P})
+    return y, Uy, state
 
-    # implementation of the state-space formulas from the paper
-    for n in range(len(y)):
-        for k in range(p):  # derivative w.r.t. a_1,...,a_p
-            dz1[:, k] = A * dz[:, k] + np.squeeze(Aabl[:, :, k]) * z
-            phi[k] = c * dz[:, k] - b0 * z[k]
-        phi[p + 1] = -np.matrix(a[1:]) * z + x[n]  # derivative w.r.t. b_0
-        for k in range(p + 2, 2 * p + 1):  # derivative w.r.t. b_1,...,b_p
-            phi[k] = z[k - (p + 1)]
-        P = A * P * A.T + noise[n] ** 2 * (bs * bs.T)
-        y[n] = c * z + b0 * x[n]
-        Uy[n] = phi.T * Uab * phi + c * P * c.T + b[0] ** 2 * noise[n] ** 2
-        z = A * z + bs * x[n]  # update of the state equations
-        dz = dz1
 
-    Uy = np.sqrt(np.abs(Uy))  # calculate point-wise standard uncertainties
+def _tf2ss(b, a):
+    """
+    Variant of :py:func:`scipy.signal.tf2ss` that fits the definitions of [Link2009]_ 
+    """
+
+    p = len(a) - 1
+    A = np.vstack([np.eye(p-1, p, k=1), -a[1:][::-1]])
+    B = np.zeros((p, 1))
+    B[-1] = 1
+    C = np.expand_dims((b[1:] - b[0] * a[1:])[::-1], axis=0)
+    D = np.ones((1,1))*b[0]
+
+    return A, B, C, D
 
-    return y, Uy
+
+def _get_derivative_A(size_A):
+    "build tensor representing dA/dtheta"
+    dA = np.zeros((size_A, size_A, size_A))
+    for k in range(size_A):
+        dA[k, -1, -(k+1)] = -1
+
+    return dA
+
+
+def _get_corr_unc(b, a, Ux):
+    """
+    Calculate the cumulated correlated noise based on equations (20) of [Link2009]_ .
+    """
+
+    # get impulse response of IIR defined by (b,a)
+    h_theta = dimpulse((b, a, 1), x0 = 0.0, t=np.arange(0, len(Ux), step=1))[1][0]
+    
+    # equation (20), note:
+    # - for values r<0 or s<0 the contribution to the sum is zero (because h_theta is zero)
+    # - Ux is the one-sided autocorrelation and assumed to be zero outside its range
+    corr_unc = np.sum(toeplitz(Ux) + toeplitz(h_theta))
+
+    return corr_unc
+
+
+def get_initial_state(b, a, Uab=None, x0 = 1.0, U0 = 1.0, Ux = None):
```

Comment:
`get_initial_state` should be included in `__all__` if it is used outside of _propagate_filter_

Reply:
:+1: